scipy/interpolate/_bsplines.py

```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+    return c
+
+def _periodic_nodes(x,l=1,r=1):
+    '''
+    returns vector of nodes on a circle
+    ``max(r, l)`` assumed to be greater than ``len(x)``
+    '''
+    dx = np.diff(x)
+    t = np.zeros(len(x) + l + r)
+    t[:l] = [x[0] - sum(dx[-i:]) for i in range(l,0,-1)]
+    t[l:-r] = x
+    t[-r:] = [x[-1] + sum(dx[:i]) for i in range(1,r+1)]
+    return t
+
+def _make_periodic_spline(x, y, t, k, axis):
+    '''
+    Compute the (coefficients of) interpolating B-spline with periodic
+    boundary conditions.
+
+    Parameters
+    ----------
+    x : array_like, shape (n,)
+        Abscissas.
+    y : array_like, shape (n,)
+        Ordinates.
+    k : int
+        B-spline degree.
+    t : array_like, shape (n + 2 * k,).
+        Nodes taken on a circle, ``k`` on the left and ``k`` on the right
+        of the vector ``x``.
+
+    Returns
+    -------
+    b : a BSpline object of the degree ``k`` and with knots ``t``.
+
+    Notes
+    -----
+    The original system is formed by ``n + k - 1`` equations where the first
+    ``k - 1`` of them stand for the ``k - 1`` derivatives continuity on the
+    edges while the other equations correspond to an interpolating case
+    (matching all the input points). Due to a special form of knot vector, it
+    can be proved that in the original system the first and last ``k``
+    coefficients of a spline function are the same, respectively. It follows
+    from the fact that all ``k-1`` derivatives are equal term by term at ends
+    and that the matrix of the original system of linear equations is
+    non-degenerate. So, we can reduce the number of equations to ``n - 1`` (first
+    ``k-1`` equations could be reduced). Another trick of this implementation is
+    cyclic shift of values of B-splines due to equality of ``k`` unknown
+    coefficients. With this we can receive matrix of the system with upper right
+    and lower left ‘blocks’, and ``k`` diagonals.  It allows to use Woodbury
+    formula to optimize the computations.
+
+    For now, this function works only for odd ``k``.
+
+    '''
+    if k % 2 == 0:
+        raise NotImplementedError("Even k periodic case is not implemented yet.")    
+    n = y.shape[0]
+
+    if n <= k:
+        raise ValueError("n should be greater than k to form system.")
+    
+    # solving periodic case using Woodbury formula
+    # set up RHS
+    A = np.zeros((k, n - 1)) # matrix of diagonals suitable for 'solve_banded'
+    for i in range(n-1):
+        A[:,i] = _bspl.evaluate_all_bspl(t, k, x[i], i + k)[:-1][::-1]
+    offset = int((k-1)/2)
+    # upper right and lower left blocks of the original matrix
+    ur = np.zeros((offset,offset))
+    ll = np.zeros((offset,offset))
+    for i in range(1,offset + 1):
+        A[offset - i] = np.roll(A[offset - i],i)
+        if k % 2 == 1 or i < offset:
+            A[offset + i] = np.roll(A[offset + i], -i)
+            ur[-i:, i - 1] = np.copy(A[offset + i, -i:])
+        ll[-i, :i] = np.copy(A[offset - i, :i])
```

Comment:
certainly need some more docs to follow this part.

Maybe also consider writing explicit loops and move them to cython, like it's done for other boundary conditions.